tools/codegen/gen.py

```diff
@@ -231,41 +231,80 @@ def get_static_dispatch_backend(f: NativeFunction, backend_index: BackendIndex)
+            output.append(f'#include <ATen/ops/{f.root_name}_{dispatch_key.lower()}_dispatch.h>')
+    return '\n'.join(output)
 
 
-def static_dispatch_extra_headers(backend: Optional[BackendIndex], skip_tensor_include: bool = False) -> List[str]:
+def static_dispatch_extra_headers(
+        backends: List[BackendIndex],
+        skip_tensor_include: bool = False
+) -> List[str]:
     if skip_tensor_include:
         # See Note [Avoiding Include Cycles In Static Dispatch]
         maybe_inl = '_inl'
     else:
         maybe_inl = ''
-    return [f'#include <ATen/{dispatch_key}Functions{maybe_inl}.h>'
-            for dispatch_key in static_dispatch_keys(backend)]
+    inl_func_headers = [f'#include <ATen/{dispatch_key}Functions{maybe_inl}.h>'
+                        for dispatch_key in static_dispatch_keys(backends)]
+    return inl_func_headers
+
+def generate_static_dispatch(
+    f: NativeFunction,
+    cpp_sig: CppSignature, *,
+    method: bool,
+    backend_index: Optional[BackendIndex]
+) -> str:
+    if backend_index is None or f.manual_kernel_registration:
+        return None
+    target_sig = CppSignatureGroup.from_native_function(f, method=False, fallback_binding=False).signature
+    name = target_sig.name()
+    exprs = translate(cpp_sig.arguments(), target_sig.arguments(), method=method)
+    exprs_str = ', '.join(a.expr for a in exprs)
+    if f.structured_delegate is not None:
+        # TODO: for ops with structured_delegate it should check the dispatch table of
+        # the out variant instead. For now, these structured ops all have CPU/CUDA kernels
+        # so we always dispatch to the `backend`, but this could be wrong when we
+        # migrate math/default_backend ops to use structured delegate.
+        if backend_index.has_kernel(f) or backend_index.dispatch_key in STRUCTURED_DISPATCH_KEYS:
+            return f'return at::{backend_index.dispatch_key.lower()}::{name}({exprs_str});'
+        else:
+            return f'TORCH_CHECK(false, "Static dispatch does not support {name} for {backend_index.dispatch_key}.");'
 
+    if backend_index.has_kernel(f):
+        return f'return at::{backend_index.dispatch_key.lower()}::{name}({exprs_str});'
+    elif f.has_composite_explicit_autograd_kernel:
+        return f'return at::{DispatchKey.CompositeExplicitAutograd.lower()}::{name}({exprs_str});'
+    elif f.has_composite_implicit_autograd_kernel:
+        return f'return at::{DispatchKey.CompositeImplicitAutograd.lower()}::{name}({exprs_str});'
+
+    return f'TORCH_CHECK(false, "Static dispatch does not support {name} for {backend_index.dispatch_key}.");'
 
 def static_dispatch(
         f: NativeFunction, cpp_sig: CppSignature,
-        *, method: bool, backend_index: Optional[BackendIndex]
+        *, method: bool, backend_indices: List[BackendIndex]
 ) -> Optional[str]:
-    if backend_index is None or f.manual_kernel_registration:
+    if backend_indices is None or f.manual_kernel_registration:
```

Comment:
you should also be able to remove this `None` check

Reply:
oh, it should probably be a check for `len(backend_indices) == 0`? And the codegen should default it to the empty list, instead of none?